internal/report/manager.go

```diff
@@ -0,0 +1,162 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"gopkg.in/yaml.v3"
+)
+
+// Manager represents a structure that manages all of the different reporters
+type Manager struct {
+	client    Client
+	reporters map[string]Reporter
+	mutex     sync.Mutex
+}
+
+// SetClient sets the client of the manager to the passed in client
+func (m *Manager) SetClient(client Client) error {
+	if client == nil {
+		return errors.New("client must not be nil")
+	}
+
+	m.client = client
+
+	return nil
+}
+
+// ResetConfig resets the current config
+func (m *Manager) ResetConfig(configData []byte) error {
+	// Create a basic map so we can unmarshal on reporter specific configs
+	cfg := make(map[string]any)
+
+	if err := yaml.Unmarshal(configData, cfg); err != nil {
+		return fmt.Errorf("failed to unmarshal config: %w", err)
+	}
+
+	// Lock so we don't have multiple configs being processed concurrently
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	// Iterate through all reporter configs and process
+	for kind, rawCfg := range cfg {
+		switch kind {
+		case snapShotKind:
+			var ssCfg snapshotConfig
+			if err := unmarshalReporterConfig(rawCfg, &ssCfg); err != nil {
+				return fmt.Errorf("failed to unmarshal Snapshot config: %w", err)
+			}
+
+			// Verify we have a snapshot reporter initialized
+			reporter, ok := m.reporters[kind]
+			if !ok {
+				reporter = NewSnapshotReporter(m.client)
+				m.reporters[kind] = reporter
+			}
+
+			// Reconfigure reporter
+			if err := m.reconfigureReporter(reporter, &ssCfg); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unrecognized reporter kind %s", kind)
+		}
+	}
+
+	return nil
+}
+
+func (m *Manager) reconfigureReporter(reporter Reporter, cfg any) error {
+	// We don't want to get stuck stopping the reporter so give it 5 seconds
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	// Stop the reporter
+	if err := reporter.Stop(ctx); err != nil {
+		return fmt.Errorf("reporter %s failed to stop: %w", reporter.Kind(), err)
+	}
+
+	// Apply the new config
+	if err := reporter.Report(cfg); err != nil {
```

Comment:
Yes, the snapshot is on demand only. So we buffer every payload that comes through trying to keep the buffer the `idealSize`. Then we report what's in the buffer when asked. Otherwise we just recycle out old items as the buffer has new items added.

The reasoning behind this is if anything ever went through the pipeline we will always have something to report. The idea is you can get a snapshot of the logs/metrics/traces that pass in or out of a certain component. Then change something and get another snapshot to see how you changes things. I think the main use case is for filtering.